### ISSUE TITLE
transformer修改超参训练后，模型infer报错

### ISSUE THREAD

**Issue by Lizhengo** [PaddlePaddle/models#2221]
paddle版本的transformer，如果修改了超参数（max_length改为64，d_model改为256，d_inner_hid改为512，n_head改成12），训练可以正常训练，但是预测时会出这个错误
<img width="1144" alt="c1e61a3e698da97918a8023e55850a53" src="https://user-images.githubusercontent.com/28192236/57600953-d079ea00-758d-11e9-8275-710c42a4ba67.png">

**Comment by guoshengCS** [comment]
这个是用的models里最新的代码吗，可否贴下while内beam_search部分的代码呢

**Comment by Lizhengo** [comment]
def beam_search():
        max_len = layers.fill_constant(
            shape=[1],
            dtype=start_tokens.dtype,
            value=max_out_len,
            force_cpu=True)
        step_idx = layers.fill_constant(
            shape=[1], dtype=start_tokens.dtype, value=0, force_cpu=True)
        cond = layers.less_than(x=step_idx, y=max_len)  # default force_cpu=True
        while_op = layers.While(cond)
        # array states will be stored for each step.
        ids = layers.array_write(
            layers.reshape(start_tokens, (-1, 1)), step_idx)
        scores = layers.array_write(init_scores, step_idx)
        # cell states will be overwrited at each step.
        # caches contains states of history steps in decoder self-attention
        # and static encoder output projections in encoder-decoder attention
        # to reduce redundant computation.
        caches = [
            {
                "k":  # for self attention
                layers.fill_constant_batch_size_like(
                    input=start_tokens,
                    shape=[-1, n_head, 0, d_key],
                    dtype=enc_output.dtype,
                    value=0),
                "v":  # for self attention
                layers.fill_constant_batch_size_like(
                    input=start_tokens,
                    shape=[-1, n_head, 0, d_value],
                    dtype=enc_output.dtype,
                    value=0),
                "static_k":  # for encoder-decoder attention
                layers.create_tensor(dtype=enc_output.dtype),
                "static_v":  # for encoder-decoder attention
                layers.create_tensor(dtype=enc_output.dtype)
            } for i in range(n_layer)
        ]

        with while_op.block():
            pre_ids = layers.array_read(array=ids, i=step_idx)
            # Since beam_search_op dosen't enforce pre_ids' shape, we can do
            # inplace reshape here which actually change the shape of pre_ids.
            pre_ids = layers.reshape(pre_ids, (-1, 1, 1), inplace=True)
            pre_scores = layers.array_read(array=scores, i=step_idx)
            # gather cell states corresponding to selected parent
            pre_src_attn_bias = layers.gather(
                trg_src_attn_bias, index=parent_idx)
            pre_pos = layers.elementwise_mul(
                x=layers.fill_constant_batch_size_like(
                    input=pre_src_attn_bias,  # cann't use lod tensor here
                    value=1,
                    shape=[-1, 1, 1],
                    dtype=pre_ids.dtype),
                y=step_idx,
                axis=0)
            logits = wrap_decoder(
                trg_vocab_size,
                max_in_len,
                n_layer,
                n_head,
                d_key,
                d_value,
                d_model,
                d_inner_hid,
                prepostprocess_dropout,
                attention_dropout,
                relu_dropout,
                preprocess_cmd,
                postprocess_cmd,
                weight_sharing,
                dec_inputs=(pre_ids, pre_pos, None, pre_src_attn_bias),
                enc_output=enc_output,
                caches=caches,
                gather_idx=parent_idx)
            # intra-beam topK
            topk_scores, topk_indices = layers.topk(
                input=layers.softmax(logits), k=beam_size)
            accu_scores = layers.elementwise_add(
                x=layers.log(topk_scores), y=pre_scores, axis=0)
            # beam_search op uses lod to differentiate branches.
            topk_indices = layers.lod_reset(topk_indices, pre_ids)
            # topK reduction across beams, also contain special handle of
            # end beams and end sentences(batch reduction)
            selected_ids, selected_scores, gather_idx = layers.beam_search(
                pre_ids=pre_ids,
                pre_scores=pre_scores,
                ids=topk_indices,
                scores=accu_scores,
                beam_size=beam_size,
                end_id=eos_idx,
                return_parent_idx=True)
            layers.increment(x=step_idx, value=1.0, in_place=True)
            # cell states(caches) have been updated in wrap_decoder,
            # only need to update beam search states here.
            layers.array_write(selected_ids, i=step_idx, array=ids)
            layers.array_write(selected_scores, i=step_idx, array=scores)
            layers.assign(gather_idx, parent_idx)
            layers.assign(pre_src_attn_bias, trg_src_attn_bias)
            length_cond = layers.less_than(x=step_idx, y=max_len)
            finish_cond = layers.logical_not(layers.is_empty(x=selected_ids))
            layers.logical_and(x=length_cond, y=finish_cond, out=cond)

        finished_ids, finished_scores = layers.beam_search_decode(
            ids, scores, beam_size=beam_size, end_id=eos_idx)
        return finished_ids, finished_scores

**Comment by Lizhengo** [comment]
使用的是还未迁移前的transformer代码

**Comment by guoshengCS** [comment]
麻烦在layers.beam_search前使用layers.Print打印下pre_ids，pre_scores，topk_indices，accu_scores这些内容贴下看下

**Comment by Lizhengo** [comment]
1557756324      The content of pre_ids:         Tensor[array_read_0.tmp_0]
        shape: [10,1,1,]
        dtype: l
        LoD: [[ 0,1,2,3,4,5,6,7,8,9,10, ][ 0,1,2,3,4,5,6,7,8,9,10, ]]
        data: 0,0,0,0,0,0,0,0,0,0,
1557756324      The content of pre_scores:      Tensor[array_read_1.tmp_0]
        shape: [10,1,]
        dtype: f
        LoD: [[ 0,1,2,3,4,5,6,7,8,9,10, ][ 0,1,2,3,4,5,6,7,8,9,10, ]]
        data: 0,0,0,0,0,0,0,0,0,0,
1557756324      The content of topk_indices:    Tensor[lod_reset_0.tmp_0]
        shape: [10,3,]
        dtype: l
        LoD: [[ 0,1,2,3,4,5,6,7,8,9,10, ][ 0,1,2,3,4,5,6,7,8,9,10, ]]
        data: 11,18,13,58,24,40,339,56,68,56,9,48,13,9,15,20,34,14,14,2,20,15,336,52,1004,33,9,89,9,246,
1557756324      The content of accu_scores:     Tensor[top_k_0.tmp_0]
        shape: [10,3,]
        dtype: f
        data: -0.358553,-2.89971,-3.73105,-2.02876,-2.83704,-3.03042,-2.5456,-2.98245,-3.13786,-1.70857,-3.0908,-3.356,-1.29264,-2.31007,-3.02268,-1.69485,-1.73231,-2.61724,-1.40763,-3.44958,-3.51778,-1.14803,-2.90526,-3.45521,-2.92069,-3.01392,-3.12794,-1.91922,-2.9766,-3.02685,

**Comment by Lizhengo** [comment]
跑完这个case后就报错了

**Comment by guoshengCS** [comment]
可以试下在layers.beam_search前加下
```python
accu_scores = layers.lod_reset(accu_scores, pre_ids)
```

**Comment by Lizhengo** [comment]
加上accu_scores = layers.lod_reset(accu_scores, pre_ids)后已解决，感谢！
